snuba/utils/streams/processing.py

```diff
@@ -30,9 +30,25 @@ class ProcessingStrategy(ABC, Generic[TPayload]):
     """
 
     @abstractmethod
-    def process(self, message: Optional[Message[TPayload]]) -> None:
+    def poll(self) -> None:
         """
-        Process a message, or periodic heartbeat (``None`` value.)
+        Poll the processor.
+
+        This method is called on each consumer loop iteration. It can be used
+        to check on the status of asynchronous tasks or perform other
+        scheduled tasks.
+        """
```

Comment:
This method is intended to be nonblocking, I'll update the docstring to reflect that.

Reply:
This is a good question and something I hadn't put a ton of thought into. My thinking is this (`close`) is called in two circumstances:

1. after the consumer process receives `SIGINT`/`SIGTERM` to prompt a shutdown, and
2. after the consumer assignment is revoked.

In either case, the processing strategy should expect to stop receiving new messages, and should reject calls to `submit` after `close` has been called. (I don't think there is really any question about that.) I don't think that necessarily means the processing strategy should discard any existing work it has already received, but at the same time, we can't wait an indefinite amount of time for work to complete or we risk overrunning the rebalancing timeout (I forget what specific configuration this is) and having the work rescheduled anyway.

This probably needs two be two separate methods on the processing strategy:

1. `close` (nonblocking), to signal that no more calls to `submit` should be expected/accepted; essentially a graceful shutdown signal,
2. `join` (blocking, with timeout), to wait for any in-flight work to complete (basically entering a blocking `poll` loop), the timeout here should ensure the maximum amount of work is completed before the rebalancing timeout is reached

This makes sense for the strategy as a whole, but gets more complex with how I had considered composing processing steps into a strategy. Using the three-step example from the description (filter, transform, collect), if `close` was to be forwarded to all three processing steps immediately when received by the processing strategy, the collect step would reject any additional items that were submitted to transform before it was closed.

It seems like it should work something like this: when the processing strategy is closed (due to one of the two reasons mentioned earlier) the consumer calls `close` and then `join` on the processing strategy, waiting for up to N seconds. Within the processing strategy, that would result in the outermost (filter) step being closed immediately, and then similarly joined for up to the timeout. As soon as the outermost step has shutdown, the subsequent step is closed (as it will receive no more items from the prior step), and joined for up to the remaining duration of the timeout. This would continue until there are no more steps left to close/shutdown, or until the timeout has elapsed.

It's possible that a very slow processing step could starve out subsequent steps, since the amount of time we're willing/able to wait for each step decreases as we move deeper into the pipeline. For example, if we had a 30 second timeout on shutdown, and filter exits immediately but transform takes 25 seconds to shut down, that leaves 5 seconds for collect to exit. I don't think this should be a problem, but worth noting.

Reply:
I think the join/close approach is fine. It is the same we do to drain a server. After a drain operation it won't receive requests anymore right away, then we hope it will wrap the requests in flight. If it does within a time limit we (shouldn't but we do) kill with -9.

Do you plan to add the join method here or in a follow up?

Reply:
I'll add it here. I also need to document what `submit` throws when it is unable to accept new messages per https://github.com/getsentry/snuba/pull/1228#discussion_r473451503 as well as a couple other minor documentation improvements.